Feat/makeup: 메이크업 추천

sdc7250 opened the pull request:
# 작업내용
- AI 로직 요청 부분 작성
- 이미지 인코딩, 디코딩 구현
- 커스터마이징 구현

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신기능
  - 메이크업 저장·목록·상세·삭제 API 제공(이미지 업로드 및 URL 응답, 페이지네이션 지원).
  - 스타일 추천, 시뮬레이션, 커스터마이즈 AI 기능 추가(업로드한 이미지 기반 결과 제공).
- 리팩터링
  - API 기본 경로를 /beautiq/makeup → /makeup 으로 변경.
  - 추천 응답 형식을 이미지 이름·URL 중심으로 개편.
  - 이미지 관리 방식 개선(임시 업로드→저장, 삭제, Base64 변환)으로 안정성 향상.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alsdn4956 committed:
feat: UserController에 회원가입·수정·조회 API 추가 및 추가 리팩토링

UserController에 회원가입·수정·조회 API 추가. username, password 필드 사용하도록 DTO와 Entity 리팩터링. UserRepository·UserService 로직 및 메서드 시그니처 개선, 보안 설정에 신규 엔드포인트 허용.

alsdn4956 committed:
feat: profileImage 수정 기능 추가

sdc7250 committed:
fix: api 명세에 맞춰 controller 수정

sdc7250 committed:
fix: api 명세에 맞춰 controller 수정

sdc7250 committed:
fix: multipartfile 최대 용량 설정

sdc7250 committed:
feat: makeup 저장 요청 dto

sdc7250 committed:
fix: User -> UserEntity로 수정

sdc7250 committed:
refactor: 이미지 추가에서 이미지 링크 추가로 수정하여 메서드명 변경

sdc7250 committed:
fix: User -> UserEntity 수정, isLiked 삭제 및 이미지url 추가로 변수 변경, SkinAnaysis 연관관계 삭제

sdc7250 committed:
fix: User -> UserEntity 수정

sdc7250 committed:
fix: 이미지이름:이미지url로 매핑되도록 수정

sdc7250 committed:
fix: analysisId 삭제

sdc7250 committed:
fix: isLiked가 없어졌으므로 관련 메서드 삭제

sdc7250 committed:
fix: imageUrl->imageName으로 수정

sdc7250 committed:
refactor: 불필요한 변수 넘기는 부분 수정 및 todo 추가

sdc7250 committed:
feat: 불필요한 메서드 삭제 및 로직 계획 주석 작성

sdc7250 committed:
fix: 이미지 업로드 반환 타입 수정 및 영구 저장 메서드 추가

sdc7250 committed:
fix: 불필요한 toString() 삭제

sdc7250 committed:
feat: 이미지 영구 저장 구현

sdc7250 committed:
fix: dto에서 Map으로 주던 "이미지이름:이미지url"을 객체 RecommendationItem으로 변경 및 RecommendationItem이 사용된 dto 수정

sdc7250 committed:
feat: 이미지 이름으로 메이크업 조회하기 추가. 이미지 이름 또한 UUID로 되어있어 중복 가능성이 낮아 실질적으로 안전해서 메이크업 조회에 imageName으로 조회하기 사용

sdc7250 committed:
fix: 반환값, api 경로 수정

sdc7250 committed:
feat: 커스터마이즈 제외 메인 로직 구현

sdc7250 committed:
feat: 요청/응답 값에 맞춘 dto 추가/변경

sdc7250 committed:
feat: 이미지 이름으로 s3 조회 및 다운로드 메서드 추가

sdc7250 committed:
feat: 이미지 이름으로 s3 조회 및 다운로드 메서드 추가

sdc7250 committed:
feat: 조회와 삭제를 쿼리 파라미터로 받도록 수정, customize 구현

sdc7250 committed:
feat: 메이크업 상세 조회 및 삭제 메서드의 파라미터를 이미지 이름으로 변경, customize 구현

sdc7250 committed:
feat: customize dto 및 공통 부분 객체 구현

sdc7250 committed:
fix: 제약조건 검증 필드 추가, NPE 방지를 위한 리스트 초기화

sdc7250 committed:
fix: Jackson이 역직렬화 할 수 있도록 기본 생성자 추가

sdc7250 committed:
fix: 사용자 ID 없이 조회, 삭제가 불가능 하도록, userId까지 사용하여 조회, 삭제하도록 시그니처 및 로직 수정

sdc7250 committed:
fix: keywords NPE 방지를 위해 수정

sdc7250 committed:
fix: s3 이미지 저장 과정에서 deleteObject에서 temp/로 시작하지 않는 이미지 삭제하는 것을 방지

sdc7250 committed:
fix: 서비스 레이어에서 소유권 검증

sdc7250 committed:
fix: 메이크업 삭제 시 s3에서 이미지도 같이 삭제, deleteImage 메서드 추가

sdc7250 committed:
fix: AI 응답이 너무 느린 경우 대비하여 비동기화

sdc7250 committed:
fix: 반환된 추천 이미지 유효 검증

sdc7250 committed:
fix: styleRecommend와 동일하게 AI 성능을 고려한 비동기화

sdc7250 committed:
fix: styleRecommend와 동일하게 AI 성능을 고려한 비동기화

sdc7250 committed:
fix: decoding 할 때 유효한 base64 검증

sdc7250 committed:
fix: 이미지 다운로드 시 이미지 이름 유효성 검증 및 메서드명 변경

sdc7250 committed:
fix: 배열 초기화와 루프 범위 통일

BORB-CHOI committed:
feat(wishlist product): 제품 추천 로직 변경으로 인한 제품 찜 목록 관리 테이블 별도 추가

BORB-CHOI committed:
refactor(product): 데이터셋 초기 업로드 기능 구현을 위한 임시 주석처리

sdc7250 committed:
refactor: 이미지 확장자를 보존하도록 수정 및 주석 내용에 파일과 이미지 혼용하던 것을 이미지로 통일

BORB-CHOI committed:
fix(AWS S3): 실제로 S3 기능을 처음 호출할 때까지 AmazonS3 객체 생성을 뒤로 미뤄서 기동 실패 방지

목적: S3 미설정 환경(로컬, 초기 개발, 테스트)에서 기동 실패 방지.
@Lazy (빈 생성 지연)
과거 문제: 애플리케이션 기동 시점에 불필요하게 AWS 자격 증명/네트워크 호출 → 지연/간헐 실패.
해결 전략: 실제로 S3 기능을 처음 호출할 때까지 AmazonS3 객체 생성을 뒤로 미룸 → 기동 안정화·속도 개선.
테스트(특히 DB 로딩이나 다른 Runner 실행) 단계에서 S3 전혀 안 쓰면 끝까지 생성 안 됨.

BORB-CHOI committed:
chore(data sets): 데이터셋 초기 로드를 위한 환경 설정.

sdc7250 committed:
fix: 이미지 입력 유효 검증 로직 추가

BORB-CHOI committed:
feat(product): 기존 찜 목록 역할을 하던 product를 데이터셋 저장을 위한 용도로 변경

BORB-CHOI committed:
feat(dataset loader): 데이터셋 초기 로드 기능 구현

sdc7250 committed:
refactor: 이미지 삭제 실패 시 로그 내용 수정

BORB-CHOI commented on src/main/java/spring/beautiq/domain/makeup/MakeUpService.java:
```diff
@@ -1,105 +1,348 @@
+        }
+        String contentType = sourceImage.getContentType();
+        if (contentType == null || !contentType.startsWith("image/")) {
+            throw new IllegalArgumentException("Invalid image file");
```
예외 처리 부분은 MakeupException 파일을 만들어 거기서 한 번에 관리하면 더욱 좋습니다. 제가 skinAnalysis와 product 부분 exception 만든 코드 참고하시면 되겠습니다.

BORB-CHOI replied:
아 혹시 이 부분 아직 안하셨다면 그냥 알고만 넘기셔도 됩니다. 시간도 촉박하고 짜피 컨벤션이라!
@sdc7250

BORB-CHOI commented on src/main/java/spring/beautiq/domain/makeup/MakeUpController.java:
```diff
@@ -1,80 +1,109 @@
 @RequiredArgsConstructor
-@RequestMapping("/beautiq/makeup")
-@MemberGuard
+@RequestMapping("/makeup")
```
URL 경로 명세서와 맞춰주시길 부탁드립니다.

BORB-CHOI commented on src/main/java/spring/beautiq/domain/makeup/dto/web/RecommendDetailResponseDto.java:
혹 시간적 여유가 되신다면, 필요하다 판단하여 명세서와 다르게 추가/수정된 부분을 명세서에 작성해주시면 감사드리겠습니다. (특히 DTO 부분이 많이 추가된 것 같습니다)

BORB-CHOI committed:
feat(skinanlaysis entity): 피부 분석 entity 네이밍 수정, 필드 구성 수정 버전 반영

BORB-CHOI committed:
feat(skinAnalysis dto): 피부 분석 dto 수정 버전 반영

BORB-CHOI committed:
fix(skinAnalysis entity): scale 불가 문제 해결

BORB-CHOI committed:
refactor(skinAnalysis dto): monthly, daily의 코드 리팩토링

BORB-CHOI committed:
fix(product dto): 테스트를 위한 임시 수정

BORB-CHOI committed:
fix(makeup/product service): 불필요 및 에러 원인 import 제거

BORB-CHOI committed:
feat(skinAnalysis controller,calculator,service): 피부 분석 변경 사항에 따른 로직 수정

BORB-CHOI committed:
feat(currentUserId): 클라이언트가 userId를 명시적으로 전송할 필요가 없고, Swagger 문서에도 숨기도록 설정

BORB-CHOI committed:
fix(YearlyDaySkinPoitsResponse): 의도되지 않은 필드 숨김 처리 해제

BORB-CHOI committed:
fix(SkinAnalysisResponse): createdAt 응답 포맷(Z) 누락으로 인한 Schema 불일치 해결

BORB-CHOI committed:
fix(SkinAnalysisResponse): 날짜 형식 불일치 수정

BORB-CHOI committed:
fix(SkinAnalysis Entity): Float에 @Min/@max는 부적절 — @DecimalMin/@DecimalMax로 교체

BORB-CHOI committed:
refactor(SkinAnalysis service): within60Days 최대 60개 보장되지 않음 — 날짜별 집계 로직 추가
                                getSixtyDayTrends에서 analyses를 날짜별로 그룹핑(예: Collectors.groupingBy)한 뒤 DayPoint 리스트를 생성해 최대 60개만 반환하도록 수정

BORB-CHOI committed:
refactor(SkinAnalysis Entity): 불필요한 디폴트 매개변숫값 대입 제거

BORB-CHOI committed:
fix(YearlyDaySkinPointsResponse): API 예제와 실제 응답 불일치 수정 (feedbackType 필드가 누락)

alsdn4956 committed:
remove: DTO 세분화

request, response로 나누었습니다

alsdn4956 committed:
feat: OIDC user support

alsdn4956 committed:
Enhance: OAuth2 success handler principal support

alsdn4956 committed:
refactor: User엔드포인트 및 세부기능

BORB-CHOI committed:
feat(product dto): 제품 추천 관련 로직 대거 변경으로 관련 dto 전부 수정. Swagger/OpenAPI 스키마·예시·유효성(0~100 등) 대폭 보강.

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product dto): AI 요청/응답 dto 수정. Swagger/OpenAPI 스키마·예시·유효성(0~100 등) 대폭 보강.

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product wishProduct dto): Wishlist Product dto 수정. Swagger/OpenAPI 스키마·예시·유효성(0~100 등) 대폭 보강.

Signed-off-by: BORB <qzsec299@naver.com>

alsdn4956 committed:
refactor: 코드리뷰 반영

BORB-CHOI committed:
refactor(productResponse): 필드명 수정, import 수정.

alsdn4956 committed:
refactor: 코드리뷰 반영

BORB-CHOI committed:
feat(product dto): service 로직에 따라 dto 수정 및 생성과 스웨거 문서 보강

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product exception): exception 케이스 추가

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product repository): 제품 추천&찜에 필요한 repository 구현

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product controller): 10/2일자로 변경된 제품추천 로직에 따라 경로 및 요청/응답 수정

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product service): 10/2일자로 변경된 제품추천 로직에 따라 전체적인 코드 수정

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
refactor/style(dto): 제품, 피부분석 내부 dto의 코드 스타일 통일화 및 중복 코드 수정

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(global-exception-handler): @PreAuthorize 사용 시 발생하는 인증/권한 예외를 GlobalExceptionHandler에서 처리하도록 예외 핸들러를 추가함.

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
fix(product): 제품 데이터 필드는 null일 수 있음

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
fix(product): entity 또는 내부 필드가 null인 경우 NPE가 발생. Product.from 메서드에 entity null 체크 및 필드별 null 처리(예외 발생 또는 디폴트 값 반환) 로직을 추가

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
refactor(exception): 예외 처리 리팩토링, 스키마 제약과 불일치 해결

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
refactor(exception): 예외 처리 패턴이 일관되게 리팩토링

Signed-off-by: BORB <qzsec299@naver.com>

BORB-CHOI committed:
feat(product service): 제품 추천 시 request로 카테고리를 받지 않고 내부적으로 계산하여 포함하도록 수정

Signed-off-by: BORB <qzsec299@naver.com>

alsdn4956 committed:
feat: ErrorCode 추가

alsdn4956 committed:
fix: userId 로직 설정 오류 수정

alsdn4956 committed:
feat: 제약 조건 추가

alsdn4956 committed:
fix: errorhandling

alsdn4956 committed:
hotfix: principalName empty

로그인 시도 -> 사용자 정보(principal)가 비어 있음 오류 해결

alsdn4956 committed:
Enhance: OAuth2 and OIDC user handling

DB 연결 오류 문제 해결

sdc7250 committed:
fix: api 명세 맞추기, id 값 조회 dto에 추가

BORB-CHOI committed:
feat(wishlist product): entity 파일명, db 테이블명 컨벤션 통일

sdc7250 committed:
feat: 페이징 로직 추가

sdc7250 committed:
refactor: 메서드 이름 명확하게 변경

sdc7250 committed:
fix: Color NPE 방어

sdc7250 committed:
refactor: 주석 삭제

sdc7250 committed:
fix: index 유효성 검증

sdc7250 committed:
fix: 이미지 이름 "temp/userId/UUID"로 되도록 수정. 이미지 소유권 검증

sdc7250 committed:
fix: 이미지 저장할 때 소유권 검증

sdc7250 committed:
refactor: 불필요한 todo 삭제

sdc7250 committed:
fix: MakeUp -> MakeUpEntity로 수정. S3Service에 병합 중 불일치 구간 수정

sdc7250 committed:
Merge branch 'develop' into feat/makeup

sdc7250 committed:
fix: 로거 중복 선언 수정

BORB-CHOI commented on src/main/java/spring/beautiq/domain/makeup/dto/ai/CustomizeAiRequestDto.java:
```diff
@@ -0,0 +1,34 @@
+    @AllArgsConstructor
+    public static class EditForAi {
+        @NotBlank
+        @Pattern(regexp = "^(skin|eye|lip|blush)$", message = "region은 skin, eye, lip, blush 중 하나여야 합니다")
```
이것도 하드 코딩 보다는 enum으로 관리하는 게 좋습니다.

BORB-CHOI commented on src/main/java/spring/beautiq/domain/makeup/MakeUpController.java:
```diff
@@ -1,80 +1,114 @@
+        return ResponseEntity.status(HttpStatus.CREATED).build();
     }
 
+    // todo: 조회, 삭제 api url을 어떻게 할까요
```
todo 주석은 이제 지우셔도 될 것 같슴다.

BORB-CHOI commented on src/main/java/spring/beautiq/domain/makeup/MakeUpService.java:
```diff
@@ -1,104 +1,370 @@
+    public RecommendResponseDto styleRecommend(
+            UUID userId,
+            MultipartFile sourceImage,
+            RecommendRequestDto recommendRequestDto // todo: 사진과 키워드 둘 중 하나만 받을지 미정
```
둘 중 하나만 받는 것으로 결정되었씁니다.

BORB-CHOI replied:
음 분기 처리를 백 서버에서 할지 AI 서버에서 할지는 고민이네요 일단 넘어갑시다~! AI 코드 보고 민우님과 셋이서 의논해보는 걸로

BORB-CHOI reviewed (approved):
LGTM
어푸
코멘트 한 번씩 읽어주세요~!
